### PR TITLE
Update Agent type for sharable agents

### PR DESCRIPTION
**Pull request by RadCod3**
## Purpose
$subject
This PR introduces reusable, dependently-typed agent definitions and `ai:Prompt` query support to `ballerina/ai`.

This pull request redesigns the agent API to make agents reusable and dependently typed, extends query handling to accept Prompt objects, and adds runtime support for structured JSON outputs together with compiler-plugin support to emit JSON-schema metadata for inferred agent return types. It also includes tests exercising structured returns and negative cases, updates mock LLM responses for deterministic structured outputs, and applies a small set of dependency/version updates.

Key changes
- Agent typing refactor: replaced the module-level AgentType enum with a union type (InferredReturnAgentType | FixedReturnAgentType). InferredReturnAgentType supports typed run signatures using typedesc<Trace|anydata> and returns the requested typedesc; FixedReturnAgentType exposes a fixed anydata run and trace API. The Agent class now embeds inferred-return behavior.
- Query expansion: agent APIs, utilities, and native bindings now accept string|ai:Prompt queries to enable templated and programmatic prompts.
- Structured-output runtime: when callers request a non-string anydata typedesc, the runtime derives a response schema, injects a JSON-only structured-output instruction into the prompt, optionally strips Markdown code fences, and binds the returned JSON to the requested typedesc (or propagates an ai:Error). Trace and plain-string behaviors are preserved.
- FunctionCallAgent/Executor updates: propagate responseSchema and Prompt-typed queries through tool-selection and execution flows; support for selecting a structured-output tool and extracting JSON final answers was added.
- Compiler-plugin improvements: replaced the previous source modification task with a JsonSchemaModificationTask that collects expected-type schemas from ai:ModelProvider.generate() and ai:Agent.run() calls and emits `@ai`:JsonSchema annotations for inferred agent return types (including nested/complex shapes and negative cases). Added analysis and source-modifier logic to attach `@ai`:AgentMetadata for custom agent definitions based on detected tools.
- Tests & mocks: new and extended tests cover structured record/array/tuple/union returns, markdown-fence stripping, trace behavior, session-memory handling, error/negative scenarios, and compiler-plugin schema generation. Mock LLM behavior was extended to return deterministic structured tool-calls and content used by tests.
- Native and build updates: native Agent.run now accepts a generic query object and receives typedescs directly; minor version bumps for Java/OpenNLP native dependency and several Ballerina dependency lockfile entries.

Impact
- Enables reusable, stronger typing for agents and structured model outputs, improving type safety and reusability.
- Broadens query expressiveness by supporting Prompt objects and templating.
- Produces compile-time JSON schema metadata for inferred agent return types, aiding validation and tooling.
- Adds test coverage and deterministic mock behavior for the new structured-output flow.

### CODE REVIEW

File: ballerina/agent.bal [hunk 1]
```diff
@@ -164,30 +176,35 @@ public isolated distinct class Agent {
...
         observe:InvokeAgentSpan span = observe:createInvokeAgentSpan(self.systemPrompt.role);
         span.addId(self.uniqueId);
         span.addSessionId(sessionId);
-        span.addInput(query);
+        span.addInput(queryString);
         string systemPrompt = getFomatedSystemPrompt(self.systemPrompt);
+
+        if td !is typedesc<string> && td !is typedesc<Trace> && td is typedesc<anydata> {
+            systemPrompt = systemPrompt + check getStructuredOutputInstruction(td);
+        }
```
[the remark this answers is not in the document]
Comment (RadCod3) [reply ballerina/agent.bal]: The leakage that was addressed in the regression test was that the previous approach injected json schema instructions as a user query which accumulated duplicate schema instructions for each `agent.run` invocation. Updated the test to make that clear bfcbffb2af89

File: ballerina/agent-utils.bal [hunk 2]
```diff
@@ -26,7 +26,7 @@ type ExecutionProgress record {|
     # Unique identifier for this execution
     string executionId;
     # Question to the agent
-    string query;
+    string|Prompt query;
```
Comment (MaryamZi) [inline ballerina/agent-utils.bal:28]: @MohamedSabthar any concerns about doing this?

File: ballerina/agent.bal [hunk 3]
```diff
@@ -101,8 +92,29 @@ public type AgentConfiguration record {|
     Credential credential?;
 |};
 
+public type AgentType InferredReturnAgentType|FixedReturnAgentType;
```
Comment (MaryamZi) [inline ballerina/agent.bal]: Let's add documentation.

File: ballerina/agent.bal [hunk 4]
```diff
@@ -164,30 +176,35 @@ public isolated distinct class Agent {
...
         );
 
         observe:InvokeAgentSpan span = observe:createInvokeAgentSpan(self.systemPrompt.role);
         span.addId(self.uniqueId);
         span.addSessionId(sessionId);
-        span.addInput(query);
+        span.addInput(queryString);
         string systemPrompt = getFomatedSystemPrompt(self.systemPrompt);
+
+        if td !is typedesc<string> && td !is typedesc<Trace> && td is typedesc<anydata> {
```
Comment (MaryamZi) [inline ballerina/agent.bal]: I think we can do

```suggestion
        if td !is typedesc<string|Trace> && td is typedesc<anydata> {
```

File: ballerina/agent.bal [hunk 5]
```diff
@@ -164,30 +176,35 @@ public isolated distinct class Agent {
...
+            systemPrompt = systemPrompt + check getStructuredOutputInstruction(td);
```
Comment (MaryamZi) [inline ballerina/agent.bal]:
```suggestion
            systemPrompt += check getStructuredOutputInstruction(td);
```

File: ballerina/agent.bal [hunk 6]
```diff
@@ -216,8 +233,16 @@ public isolated distinct class Agent {
                     endTime: time:utcNow(),
                     output: {role: ASSISTANT, content: answer},
                     toolCalls
-                }
-                : answer;
+                };
+                return trace;
```
Comment (MaryamZi) [inline ballerina/agent.bal]: Can directly return without introducing another variable?

File: ballerina/agent.bal [hunk 7]
```diff
@@ -237,10 +262,55 @@ public isolated distinct class Agent {
...
+    }
+
+}
+
+# Builds an instruction, appended to the system prompt, that guides the agent to produce its final
+# answer as a JSON value conforming to the schema of the expected return type.
+#
+# + td - Type descriptor specifying the expected return type
+# + return - The instruction text, or an error if a schema cannot be derived for the type
+isolated function getStructuredOutputInstruction(typedesc<anydata> td) returns string|Error {
```
Comment (MaryamZi) [inline ballerina/agent.bal]: Is the prompt significantly different to what we are using with generate? Just wondering if we can reuse in some way.

File: ballerina/agent.bal [hunk 8]
```diff
@@ -237,10 +262,55 @@ public isolated distinct class Agent {
...
+        "schema. Respond with only the JSON value, without any additional text:\n" +
+        schema.toJsonString();
+}
+
+# Parses the agent's final answer into a value of the expected type.
+#
+# + answer - The agent's final answer (expected to be a JSON value)
+# + td - Type descriptor specifying the expected return type
+# + return - The bound value, or an error if the answer cannot be parsed into the type
+isolated function parseAnswerAsType(string answer, typedesc<anydata> td) returns anydata|Error {
```
Comment (MaryamZi) [inline ballerina/agent.bal:513]: This also I think we should be able to reuse/build on what we use with `generate`.

File: compiler-plugin-tests/src/test/resources/ballerina_sources/json_schema_tests/01_agent_run_record_return/main.bal
Comment (MaryamZi) [inline compiler-plugin-tests/src/test/resources/ballerina_sources/json_schema_tests/01_agent_run_record_return/main.bal:1]: Let's add license headers.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/ai/plugin/JsonSchemaModificationTask.java [hunk 9]
```diff
@@ -229,16 +243,30 @@ public GenerateMethodJsonSchemaGenerator(SemanticModel semanticModel, Optional<S
...
+            if (this.modelProviderSymbol == null && this.agentTypeSymbol == null) {
                 return;
             }
             visit(modulePartNode);
         }
 
         public void visit(RemoteMethodCallActionNode remoteMethodCallActionNode) {
             SimpleNameReferenceNode methodName = remoteMethodCallActionNode.methodName();
-            if (!methodName.name().text().equals(GENERATE_METHOD_NAME)) {
+            if (this.modelProviderSymbol == null || !methodName.name().text().equals(GENERATE_METHOD_NAME)) {
```
Comment (MaryamZi) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/ai/plugin/JsonSchemaModificationTask.java]: Why do we need these null checks?

Comment (MaryamZi) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/ai/plugin/JsonSchemaModificationTask.java]: These ideally cannot be null, right?

File: ballerina-tests/tests/agent-structured-return-test.bal
Comment (MaryamZi) [inline ballerina-tests/tests/agent-structured-return-test.bal:1]: Shall we add some negative tests too? Compile-time errors for trying to implement FixedReturnAgentType with a non-anydata return type in run?

Comment (RadCod3) [reply ballerina-tests/tests/agent-structured-return-test.bal:1]: Added a negative test case in 7ab3268a2bb1

File: compiler-plugin-tests/src/test/resources/ballerina_sources/json_schema_tests/01_agent_run_record_return/main.bal
Comment (RadCod3) [reply compiler-plugin-tests/src/test/resources/ballerina_sources/json_schema_tests/01_agent_run_record_return/main.bal:1]: Addressed in a3e1257e6401

File: ballerina/agent.bal (hunk 6 above)
Comment (RadCod3) [reply ballerina/agent.bal]: Addressed in a3e1257e6401

File: ballerina/agent.bal (hunk 3 above)
Comment (RadCod3) [reply ballerina/agent.bal]: Addressed in a3e1257e6401

File: ballerina/agent.bal (hunk 7 above)
Comment (RadCod3) [reply ballerina/agent.bal]: `generate` does not use a prompt, it instead forces structured output via a tool call in https://github.com/ballerina-platform/module-ballerina-ai/blob/2449680a29e67be9556b1fc63c52cc224e577eb6/ballerina/default-model-provider-utils.bal#L102

File: ballerina/agent.bal (hunk 8 above)
Comment (RadCod3) [reply ballerina/agent.bal:513]: Since `generate` is a forced tool call, it always returns a valid json. In this implementation since we are using a prompt based approach to generate the data, we get a string response from the agent, on which we use `fromJsonStringWithType` and strip markdown fences the model _may_ add. The json schema generation step reuses the same helper function/s.

File: ballerina/agent.bal (hunk 7 above)
Comment (MaryamZi) [reply ballerina/agent.bal]: Forgot :) So for `generate` (or natural expressions rather) we switched to that because tool calling and structured output were said to ensure better adherence with JSON schema rather than this "JSON mode" (relying on the prompt). I think the same should be true for agents. Shall we look into this a bit and see what the best way to do this is?

File: ballerina/agent-utils.bal (hunk 2 above)
Comment (MohamedSabthar) [reply ballerina/agent-utils.bal:28]: if we go ahead with this change, we'll likely need to handle multimodal inputs to the agent as well

File: ballerina/function-call-agent.bal [hunk 10]
```diff
@@ -106,13 +106,26 @@ isolated distinct class FunctionCallAgent {
                 availableTools = filteredTools.toString()
         );
 
+        ResponseSchema? responseSchema = progress.responseSchema;
+        if responseSchema is ResponseSchema {
+            filteredTools.push(getStructuredOutputTool(responseSchema.schema));
+        }
```
[the remark this answers is not in the document]
Comment (RadCod3) [reply ballerina/function-call-agent.bal]: The internal tool name was changed to `__ballerina_ai_structured_result__` making it highly unlikely that the tool collides with a user defined tool.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/ai/plugin/JsonSchemaModificationTask.java (hunk 9 above)
Comment (RadCod3) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/ai/plugin/JsonSchemaModificationTask.java]: Addressed in 1f405f3

**Review by MohamedSabthar — APPROVED** [review]
